docs/core/testing/unit-testing-best-practices.md

```diff
@@ -0,0 +1,515 @@
+
+*Stub* - A stub is a controllable replacement for an existing dependency (or collaborator) in the system. By using a stub, you can test your code without dealing with the dependency directly. By default, a fake starts out as a stub.
+
+Consider the following code snippet:
+
+```csharp
+var mockOrder = new MockOrder();
+var purchase = new Purchase(mockOrder);
+
+purchase.ValidateOrders();
+
+Assert.True(purchase.CanBeShipped);
+```
+
+This would be an example of Mock being used improperly. In this case, it is a stub. We're just passing in the Order as a means to be able to instantiate `Purchase` (the system under test). The name `MockOrder` is also very misleading because again, the order is not a mock.
+
+A better approach would be
+
+```csharp
+var stubOrder = new FakeOrder();
+var purchase = new Purchase(stubOrder);
+
+purchase.ValidateOrders();
+
+Assert.True(purchase.CanBeShipped);
+```
+
+By renaming the class to `FakeOrder`, we've made the class a lot more generic, the class can be used as a mock or a stub. Whichever is better for the test case. In the above example, `FakeOrder` is used as a stub. We're not using the `FakeOrder` in any shape or form during the assert. We just passed it into the `Purchase` class to satisfy the requirements of the constructor.
+
+To use it as a Mock, we could do something like this
+
+```csharp
+var mockOrder = new FakeOrder();
+var purchase = new Purchase(mockOrder);
+
+purchase.ValidateOrders();
+
+Assert.True(mockOrder.Validated);
+```
+
+In this case, we are checking a property on the Fake (asserting against it), so in the above code snippet, the `mockOrder` is a Mock.
+
+**It's important to get this terminology correct. If you call your stubs "mocks", other developers are going to make false assumptions about your intent.**
+
+The main thing to remember about mocks versus stubs is that mocks are just like stubs, but you assert against the mock object, whereas you do not assert against a stub.
+
+## Best Practices
+
+### Arranging Your Tests
+Arrange, Act, Assert is a common pattern when unit testing. As the name implies, it consists of three main actions:
+- *Arrange* your objects, creating and setting them up as necessary.
+- *Act* on an object.
+- *Assert* that something is as expected.
+
+#### Why?
+- Clearly separates what is being tested from the *arrange* and *assert* steps.
+- Less chance to intermix assertions with "Act" code.
+
+#### Bad:
+```csharp
+public void IsValidWord_InputIsNull_ReturnsFalse()
+{
+    // Arrange
+    var glossary = new Glossary();
+
+    // Assert
+    Assert.False(glossary.IsValidWord(null));
+}
+```
+
+#### Better:
+```csharp
+public void IsValidWord_InputIsNull_ReturnsFalse()
+{
+    var glossary = new Glossary();
+    
+    var result = glossary.IsValidWord(null);
+
+    Assert.False(result);
+}
+```
+
+### Naming Your Tests
+The name of your test should consist of three parts:
```

Comment:
I'd like to see this augmented with guidance for using theories (xUnit terminology) or data driven tests (MSTest term).

Also, I'd modify this guidance applies to the fully qualified name (testNamespace.TestClass.TestMethod).